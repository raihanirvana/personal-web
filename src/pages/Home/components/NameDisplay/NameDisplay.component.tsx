import React from 'react';
import { useTypewriter } from 'react-simple-typewriter';

import { Name, UseShowSecondText } from './NameDisplay.component.type';

/**
 * hooks for show second text and get typewriter props
 * 
 * @param {string} firstName first name
 * @param {string} secondName second name
 * @returns {UseShowSecondText} flag for show second text
 */
const useTypeWriterText = (firstName: string, secondName: string): UseShowSecondText => {
  const [showSecondText, setShowSecondText] = React.useState<boolean>(false);

  const [typedFirstName] = useTypewriter({
    words: [firstName],
    loop: 1,
    typeSpeed: 200,
    onLoopDone: () => { setTimeout(() => setShowSecondText(true), 500) },
  });

  const [typedSecondName] = useTypewriter({ words: [secondName], loop: 1, typeSpeed: 250, });

  return { showSecondText, typedFirstName, typedSecondName };
}


/**
 * Component to display two names sequentially using a typewriter effect.
 *
 * @param {string} props.firstName - The first name to be displayed.
 * @param {string} props.secondName - The second name to be displayed after the first one.
 * @returns {JSX.Element} - A React component rendering the typewriter effect for both names.
 */
const NameDisplay = ({ firstName, secondName }: Name): JSX.Element => {
  const { showSecondText, typedFirstName, typedSecondName } = useTypeWriterText(firstName, secondName)

  return (
    <div className="text-white">
      <p className="text-5xl font-rigid-reguler">{typedFirstName}</p>
      {showSecondText && (
        <p className="text-9xl font-rigid-medium">{typedSecondName}</p>
      )}
    </div>
  );
};

export default NameDisplay;
