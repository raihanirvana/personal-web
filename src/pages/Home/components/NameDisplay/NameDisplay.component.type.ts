export type Name = {
  firstName: string,
  secondName: string,
};

export type UseShowSecondText = {
  showSecondText: boolean,
  typedFirstName: string,
  typedSecondName: string
}