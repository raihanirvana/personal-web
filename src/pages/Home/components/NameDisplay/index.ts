export { default } from './NameDisplay.component'