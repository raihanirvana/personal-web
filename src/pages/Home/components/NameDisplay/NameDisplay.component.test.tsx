import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import NameDisplay from './NameDisplay.component';

jest.mock('react-simple-typewriter', () => ({
  useTypewriter: ({ words, onLoopDone }: { words: string[], onLoopDone?: () => void }) => {
    React.useEffect(() => {
      if (onLoopDone) {
        setTimeout(onLoopDone, 10);
      }
    }, [onLoopDone]);

    return [words[0]];
  }
}));

describe('NameDisplay Component', () => {
  it('renders correctly after delay and matches snapshot', async () => {
    const { asFragment } = render(<NameDisplay firstName="Raihan" secondName="Ganteng" />);
  
    await waitFor(() => expect(screen.getByText('Ganteng')).toBeInTheDocument(), { timeout: 1000 });
  
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays first name immediately', () => {
    render(<NameDisplay firstName="Raihan" secondName="Raihan" />);

    expect(screen.getByText('Raihan')).toBeInTheDocument();
  });

  it('displays second name after a delay', async () => {
    render(<NameDisplay firstName="Raihan" secondName="Ganteng" />);

    await waitFor(() => expect(screen.getByText('Ganteng')).toBeInTheDocument(), { timeout: 1000 });
  });
});
