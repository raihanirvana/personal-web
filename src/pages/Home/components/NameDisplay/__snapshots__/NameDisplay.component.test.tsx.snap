// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameDisplay Component renders correctly after delay and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="text-white"
  >
    <p
      class="text-5xl font-rigid-reguler"
    >
      Raihan
    </p>
    <p
      class="text-9xl font-rigid-medium"
    >
      Ganteng
    </p>
  </div>
</DocumentFragment>
`;
